### PR TITLE
random zap amounts

### LINKED ISSUE #1112: Random zap setting
Add a setting that when enabled allows someone to zap a random amount within a specified range.

This will require adding new user settings and modifying zaps to be random when enabled.

### PR DESCRIPTION
**Pull request by SatsAllDay**
## Description
Closes #1112

Adds an option to enable random zap amounts as a user setting.

Users can opt in to random zapping in their settings. If they enable the setting, they must pick a minimum and maximum random zap in their user settings before being able to save this option.

Random zaps take precedence over a default zap value.

Random zap config is validated to ensure that a min and max is specified, and that the min is less than the max, if the feature is enabled.

## Additional Context
N/A

## Checklist

**Are your changes backwards compatible? Please answer below:**
Yes

**Did you QA this? Could we deploy this straight to production? Please answer below:**
Yes

**For frontend changes: Tested on mobile? Please answer below:**
I made the desktop window smaller but that's about it.

**Did you introduce any new environment variables? If so, call them out explicitly here:**
No.

### CODE REVIEW

File: api/typeDefs/user.js [hunk 1]
```diff
@@ -98,6 +98,9 @@ export default gql`
     noteItemMentions: Boolean!
     nsfwMode: Boolean!
     tipDefault: Int!
+    tipRandom: Boolean!
```
Comment (SatsAllDay) [inline api/typeDefs/user.js:101]: I named everything `tipRandom...`, following the existing pattern from `tipDefault` and `turboTipping`. Would you prefer me to use `zap` instead? e.g. `zapRandom`, `zapRandomMin`, etc.? @huumn

Comment (huumn) [reply api/typeDefs/user.js:101]: tip is fine. We've been so lazy with the names

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-name: Eslint Check
+name: Lint Check
```
Comment (SatsAllDay) [inline .github/workflows/lint.yml:1]: Just a nit because we don't actually use eslint.

File: components/item-act.js [hunk 3]
```diff
@@ -101,7 +101,7 @@ export default function ItemAct ({ onClose, item, down, children, abortSignal })
   return (
     <Form
       initial={{
-        amount: me?.privates?.tipDefault || defaultTips[0],
+        amount: defaultTipIncludingRandom(me?.privates) || defaultTips[0],
```
Comment (SatsAllDay) [inline components/item-act.js:104]: Honor random zap settings here, too.

File: components/upvote.js [hunk 4]
```diff
@@ -67,11 +67,20 @@ export function DropdownItemUpVote ({ item }) {
   )
 }
 
-export const nextTip = (meSats, { tipDefault, turboTipping }) => {
+export const defaultTipIncludingRandom = ({ tipDefault, tipRandom, tipRandomMin, tipRandomMax }) =>
```
Comment (SatsAllDay) [inline components/upvote.js:70]: A reusable function for determining the next zap value that honors `tipRandom` and `tipDefault`

File: lib/validate.js [hunk 5]
```diff
@@ -540,8 +540,31 @@ export const actSchema = object({
   act: string().required('required').oneOf(['TIP', 'DONT_LIKE_THIS'])
 })
 
-export const settingsSchema = object({
+export const settingsSchema = object().shape({
```
Comment (SatsAllDay) [inline lib/validate.js:543]: Had to do this for the cyclic dependency issue noted below

**Comment by SatsAllDay** [comment]
I think this is ready for review. Let me know of any feedback!

File: prisma/schema.prisma [hunk 6]
```diff
@@ -30,6 +30,9 @@ model User {
   apiKeyHash                String?              @unique(map: "users.apikeyhash_unique") @db.Char(64)
   apiKeyEnabled             Boolean              @default(false)
   tipDefault                Int                  @default(100)
+  tipRandom                 Boolean              @default(false)
+  tipRandomMin              Int?
+  tipRandomMax              Int?
```
Comment (huumn) [inline prisma/schema.prisma:35]: I don't think we need the boolean if we can just check for null on min and max. I'd suggest we use a postgres [range type](https://www.postgresql.org/docs/current/rangetypes.html) too but [prisma doesn't support it](https://github.com/prisma/prisma/issues/3287).

**Review by huumn — CHANGES_REQUESTED** [review]
Sorry for the slow review. The only thing blocking is the sort of redundant boolean. Not a huge deal, but we already have so many columns.

(I'll want to clean up the UI of the zap accordion too, but given that involves zap undos too, I'll do that after merging.)

**Comment by SatsAllDay** [comment]
> Sorry for the slow review. The only thing blocking is the sort of redundant boolean. Not a huge deal, but we already have so many columns.
> 
> 
> 
> (I'll want to clean up the UI of the zap accordion too, but given that involves zap undos too, I'll do that after merging.)

Ok, thank you for the feedback! FWIW, I went with the redundant Boolean so users could explicitly enable/disable the feature which I felt was a more clear pattern, and also so that users could disable the feature without wiping out their configured values.

In any case, I can make the adjustments, hopefully in the next day or two :)

**Comment by huumn** [comment]
I think the boolean in the UI does make sense. For the db, I'd just prefer the values. Toggling off an on wouldn't persist old values but I don't suspect that'd happen a lot.

**Comment by SatsAllDay** [comment]
What are your thoughts on keeping `tipRandom: Boolean!` as part of the `UserPrivates` in the GraphQL query API, and just having a manual resolver in `api/resolvers/user` which does the null check on `tipRandomMax` and `tipRandomMin` to resolve it, but still removing it from `SettingsInput` and the db? Basically leave it as a read-only field, computed in the API. I find it nice to have `me.privates.tipRandom` in the API response to consume in various spots in the UI.

I'm ok either way, but this idea occurred to me as I am working on the above feedback and wanted to get your take.

**Comment by SatsAllDay** [comment]
OK, I think we're ready for another review pass!

File: components/upvote.js [hunk 7]
```diff
@@ -67,11 +67,20 @@ export function DropdownItemUpVote ({ item }) {
...
   // what should our next tip be?
-  if (!turboTipping) return (tipDefault || 1)
+  const calculatedDefault = defaultTipIncludingRandom({ tipDefault, tipRandom, tipRandomMin, tipRandomMax })
+
+  if (!turboTipping) {
+    return calculatedDefault
+  }
 
-  let sats = tipDefault || 1
+  let sats = calculatedDefault
```
Comment (huumn) [inline components/upvote.js:83]: Ohhh cool compatible with turbo zapping!

**Review by huumn — APPROVED** [review]
Still need to QA but no objections otherwise.

**Pull request merged**